src/lsm/compaction.zig

```diff
@@ -34,616 +34,749 @@ const std = @import("std");
+                &compaction.iter_tracer_slot,
+                .{ .tree_compaction_iter = .{
+                    .tree_name = compaction.tree_name,
+                    .level_b = compaction.context.level_b,
+                } },
+                @src(),
             );
 
-            // Release the table's block addresses in the Grid as it will be made invisible.
-            // This is safe; iterator_b makes a copy of the block before calling us.
-            const grid = compaction.grid;
-            for (Table.index_data_addresses_used(index_block)) |address| {
-                grid.release(address);
-            }
-            for (Table.index_filter_addresses_used(index_block)) |address| {
-                grid.release(address);
-            }
-            grid.release(Table.index_block_address(index_block));
+            compaction.iter_check(.a);
         }
 
-        pub fn compact_tick(compaction: *Compaction, callback: Callback) void {
-            assert(compaction.status == .processing);
-            assert(compaction.callback == null);
-            assert(compaction.io_pending == 0);
-            assert(!compaction.merge_done);
-
-            switch (compaction.strategy) {
-                .merge => {
-                    compaction.callback = callback;
-
-                    tracer.start(
-                        &compaction.tracer_slot,
-                        .{ .tree_compaction_tick = .{
-                            .tree_name = compaction.tree_name,
-                            .level_b = compaction.level_b,
-                        } },
-                        @src(),
-                    );
-
-                    // Generate fake IO to make sure io_pending doesn't reach zero multiple times
-                    // from IO being completed inline down below. The fake IO is immediately
-                    // resolved and triggers the cpu_merge_start if all IO completes inline or if no
-                    // IO was started.
-                    compaction.io_start();
-                    defer compaction.io_finish();
-
-                    // Start reading blocks from the iterators to merge them.
-                    if (compaction.iterator_a.tick()) compaction.io_start();
-                    if (compaction.iterator_b.tick()) compaction.io_start();
-
-                    // Start writing blocks made by merge_iterator from a previous compact_tick().
-                    compaction.io_write_start(.data);
-                    compaction.io_write_start(.filter);
-                    compaction.io_write_start(.index);
-                },
-                .move => {
-                    const snapshot_max = snapshot_max_for_table_input(compaction.op_min);
-                    const level_b = compaction.level_b;
-                    const level_a = level_b - 1;
+        /// If `values_in[index]` is empty and more values are available, read them.
+        fn iter_check(compaction: *Compaction, side: Side) void {
```

Comment:
It's going to be a bit weird because the iterators take Table at comptime, but we want only have one compaction worth of allocated state. So I'm thinking about making ReadPipeline and WritePipeline objects that own blocks and have read methods. Then those get passed to LevelIndexIterator and TabelDataIterator and they use them to prefetch.